sale_commission_product_criteria/models/commission.py

```diff
@@ -0,0 +1,299 @@
+                            decimal_places,
+                        ),
+                        item.currency_id.symbol or "",
+                    )
+                else:
+                    item.commission_value = "%s %s" % (
+                        item.currency_id.symbol or "",
+                        float_repr(
+                            item.fixed_amount,
+                            decimal_places,
+                        ),
+                    )
+            elif item.commission_type == "percentage":
+                item.commission_value = str(item.percent_amount) + " %"
+
+    @api.constrains("product_id", "product_tmpl_id", "categ_id")
+    def _check_product_consistency(self):
+        for item in self:
+            if item.applied_on == "2_product_category" and not item.categ_id:
+                raise ValidationError(
+                    _(
+                        "Please specify the category for which this rule should "
+                        "be applied"
+                    )
+                )
+            elif item.applied_on == "1_product" and not item.product_tmpl_id:
+                raise ValidationError(
+                    _(
+                        "Please specify the product for which this rule should "
+                        "be applied"
+                    )
+                )
+            elif item.applied_on == "0_product_variant" and not item.product_id:
+                raise ValidationError(
+                    _(
+                        "Please specify the product variant for "
+                        "which this rule should be applied"
+                    )
+                )
+
+    @api.onchange("product_id")
+    def _onchange_product_id(self):
+        has_product_id = self.filtered("product_id")
+        for item in has_product_id:
+            item.product_tmpl_id = item.product_id.product_tmpl_id
+        if self.env.context.get("default_applied_on", False) == "1_product":
+            # If a product variant is specified, apply on variants instead
+            # Reset if product variant is removed
+            has_product_id.update({"applied_on": "0_product_variant"})
+            (self - has_product_id).update({"applied_on": "1_product"})
+
+    @api.onchange("product_tmpl_id")
+    def _onchange_product_tmpl_id(self):
+        has_tmpl_id = self.filtered("product_tmpl_id")
+        for item in has_tmpl_id:
+            if (
+                item.product_id
+                and item.product_id.product_tmpl_id != item.product_tmpl_id
+            ):
+                item.product_id = None
+
+    @api.model
+    def create(self, values):
+        values = self.validate_values(values)
+        return super(CommissionItem, self).create(values)
+
+    def write(self, values):
+        values = self.validate_values(values)
+        res = super(CommissionItem, self).write(values)
+        self.invalidate_cache()
+        return res
+
+    def validate_values(self, values):
```

Comment:
Why not doing this in a method api.constrains?

Reply:
In this method there is data modifications and  api.constrains designed only for checks. Or am i wrong ?